scipy/stats/_sensitivity_analysis.py

```diff
@@ -0,0 +1,466 @@
+    -------
+    s, st : array_like (x, d)
+        First order and total order Sobol' indices.
+
+    References
+    ----------
+    .. [1] Saltelli, A., P. Annoni, I. Azzini, F. Campolongo, M. Ratto, and
+       S. Tarantola. "Variance based sensitivity analysis of model
+       output. Design and estimator for the total sensitivity index."
+       Computer Physics Communications, 181(2):259-270,
+       :doi:`10.1016/j.cpc.2009.09.018`, 2010.
+    """
+    f_AB = f_AB.reshape(-1, *f_A.shape)
+
+    var = np.var(np.vstack([f_A, f_B]), axis=0)
+
+    s = np.mean(f_B * (f_AB - f_A), axis=1) / var
+    st = 0.5 * np.mean((f_A - f_AB) ** 2, axis=1) / var
+
+    return s.T, st.T
+
+
+@dataclass
+class SobolResult:
+    first_order: np.ndarray
+    total_order: np.ndarray
+    _A: np.ndarray
+    _B: np.ndarray
+    _AB: np.ndarray
+    _f_A: np.ndarray
+    _f_B: np.ndarray
+    _f_AB: np.ndarray
+    _bootstrap_result: BootstrapResult = None  # type: ignore[valid-type]
+
+    def bootstrap(
+        self,
+        confidence_level: DecimalNumber = 0.95,
+        n_resamples: IntNumber = 99
+    ) -> BootstrapResult:  # type: ignore[valid-type]
+        """Bootstrap Sobol' indices to provide confidence intervals.
+
+        Parameters
+        ----------
+        confidence_level : float, default: ``0.95``
+            The confidence level of the confidence intervals.
+        n_resamples : int, default: ``99``
+            The number of resamples performed to form the bootstrap
+            distribution of the indices.
+
+        Returns
+        -------
+        res : BootstrapResult
+            Bootstrap result containing the confidence intervals and the
+            bootstrap distribution of the indices.
+
+        """
+        def statistic(idx):
+            f_A_ = self._f_A[idx]
+            f_B_ = self._f_B[idx]
+
+            n, d = self._A.shape
+            f_AB_ = np.empty((d, *f_A_.shape))
+            for i in range(d):
+                f_AB_[i] = self._f_AB[np.array(idx)+i*n]
+            f_AB_ = f_AB_.reshape(-1, f_A_.shape[1])
+
+            return saltelli_2010(f_A_, f_B_, f_AB_)
+
+        n = len(self._f_A)
+
+        self._bootstrap_result = bootstrap(
+            [np.arange(n)], statistic=statistic, method="BCa",
+            batch=int(n*0.7), n_resamples=n_resamples,
+            confidence_level=confidence_level,
+            bootstrap_result=self._bootstrap_result
+        )
+
+        return self._bootstrap_result
+
+
+def sobol_indices(
+    *,
+    func: Callable,
+    n: IntNumber,
+    dists: List[PINVDist],
+    random_state: SeedType = None
+) -> SobolResult:
+    r"""Global sensitivity indices of Sobol'.
+
+    Parameters
+    ----------
+    func : callable
+        Function to compute the Sobol' indices from. It's signature must be
+        ``func(x: ArrayLike) -> ArrayLike``, with ``x`` of shape ``(-1, d)``
+        and the output should have a shape ``(-1, s)`` with ``s`` the number
+        of output.
```

Comment:
Let's make a vote for the shape of the output. At the moment it is (-1, d). Arguments can be made for both. Please vote with emojis.

* (-1, d) 🚀 
* (d, -1) 👍

Reply:
The output will have only one non-singleton dimensions, right? Assuming `x` means `x.shape` and `-1` means number of samples, these output shape options don't seem to be correct.

Reply:
`x` was meant to be a placeholder for `s, d`. What is wrong with the shapes?

Reply:
I misunderstood because `x` is also the name of a variable. I suppose the suggestion is to be explicit so that we all understand what the options are.

---

TLDR:
If the most common use case is for functions with only a few variables, I would prefer Option 3 (`n_inputs` separate inputs of shape `(n_samples,)` -> one output of shape `(n_outputs, n_samples)`), since there is already precedent for this in `stats`, and `sobol` is supposed to go in `stats`. I can see Option 2 (`(n_inputs, n_samples)` -> `(n_outputs, n_samples)`) if this function will commonly be used for functions with many inputs. 

---

There are actually several other options that come to mind besides those listed above. Here are the ones that stand out to me.

1. In `scipy.stats`, multivariate distributions and QMC engines tend to work with arrays like `(n_samples, dimensionality)`, so from this perspective, it makes sense for `func` to accept inputs of shape `(n_samples, n_inputs)` as the documentation quoted above suggests. But this leads to two options for the output shape.
    a. `(n_samples, n_outputs)`, as the documentation quoted above suggests. I suppose the thought is that `n_samples` maintains its spot along the first axis, but I don't know if there is any precedent for this elsewhere in SciPy.
    b. `(n_outputs, n_samples)`. It is really not so strange for input and output variables to go along different axes (think about a Jacobian matrix). Also, consider that this is like "consuming" the axis that corresponds with various inputs. Precedent for this option includes multivariate distribution functions and [`optimize.qmc_quad`](https://scipy.github.io/devdocs/reference/generated/scipy.integrate.qmc_quad.html#scipy.integrate.qmc_quad)).
2. [`optimize.differential_evolution`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.differential_evolution.html#scipy.optimize.differential_evolution) expects `func` to accept arrays of shapes like `(n_inputs, n_samples)` and return an array with shape `(n_samples,)` For this function, the more general output shape would be `(n_outputs, n_samples)`. 
3. [`stats.bootstrap`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.bootstrap.html#scipy.stats.bootstrap) and [`stats.permutation_test`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.permutation_test.html#scipy.stats.permutation_test
) expect a statistic `n_inputs` separate arrays, each of shape `(n_samples,)`, and the output for n-d statistics is `(n_ouputs, n_samples)`.

An advantage of 1b, 2, and 3 is that the output of functions with only one output variable (e.g. Ishigami, any multivariate distribution PDF, likelihood functions) can be one-dimensional. I think this is very important; we shouldn't require the user to return a column array (with a singleton dimension) for 1D functions. For this reason, please let's go with one of options 1b, 2, or 3; this would eliminate that `reshape` at the end of `f_ishigami` (and a lot of user code). (Yes, it might mean a reshape in `sobol`, but isn't that better than having the reshape in all the example functions?)

An advantage of 2 is that the input array can easily be unpacked into separate variables. This would simplify Ishigami from:

```python3
def f_ishigami(x)
    x = np.atleast_2d(x)
    f_eval = (
        np.sin(x[:, 0])
        + 7 * np.sin(x[:, 1])**2
        + 0.1 * (x[:, 2]**4) * np.sin(x[:, 0])
    )
    return f_eval.reshape(-1, 1)
```
to:
```python3
def f_ishigami(x)
    f_eval = (
        np.sin(x[0])
        + 7 * np.sin(x[1])**2
        + 0.1 * (x[2]**4) * np.sin(x[0])
    )
```

Option 3 takes it a step further:
```python3
def f_ishigami(x0, x1, x2):
    f_eval = (
        np.sin(x0)
        + 7 * np.sin(x1)**2
        + 0.1 * (x2**4) * np.sin(x0)
    )
```
The disadvantage here is that if the user wants to treat all the variables together in a single array, they have to pack it into an array again, and this requires copying data.

I might be regretting the choice of 1b in `qmc_quad`. I had multivariate distribution PDFs in mind when I wrote it, but it might have been more convenient for users had I gone with Option 2.

Reply:
Ok I tried to update the vote. Feel free to edit if I made a mistake.

My preference goes for 1.b or 2 if we don't want 1.a.